assets/src/data/eventespresso/filter-state/index.js

```diff
@@ -0,0 +1,53 @@
+/**
+ * External dependencies
+ */
+import { Map } from 'immutable';
+import createSelector from 'rememo';
+import { registerStore } from '@wordpress/data';
+
+const DEFAULT_STATE = {};
+const SET_FILTER = 'SET_FILTER';
+
+const getState = ( state, context ) => {
+	return typeof state[ context ] === 'undefined' ?
+		Map() :
+		state[ context ];
+};
+
+const filterStateReducer = ( state = DEFAULT_STATE, action ) => {
+	const { type, context, filter, value } = action;
+	if ( type === SET_FILTER ) {
+		return {
+			...state,
+			[ context ]: getState( state, context ).set( filter, value ),
+		};
+	}
+	return state;
+};
+
+const setFilter = ( context, filter, value ) => {
+	return {
+		type: SET_FILTER,
+		context,
+		filter,
+		value,
```

Comment:
You need to be wary about what is set as the value here. If it's an object or array, then that means that there will be a new immutable map created every time it's set even if there is a value equivalency. For instance, if this is called with something like:

```js
dispatch( 'eventespresso/filter-actions' ).setFilter( 'some-context', 'filter-name', [ 10, 20 ] );
```

And there is already a value for that filter equalling `[ 10, 20 ]` in the state, then when the reducer is called it will result in a new immutable map.  You can see a demo of this if you go to the [immutable.js](https://immutable-js.github.io/immutable-js/) site, open up your browser console, and you can do the following in the console:

```js
valueTest = Immutable.Map().set( 'foo', [ 10, 20 ] );
valueTestB = valueTest.set( 'foo', [ 10, 20 ] );
valueTest === valueTestB
> false
valueTest = ImmutableMap().set( 'foo', 10 );
valueTestB = valueTest.set( 'foo', 10 );
valueTest === valueTestB
> true
```

Of course if a reference is passed around, it wouldn't be an issue (and likely in most use-cases a reference _would_ be passed around).  However, it's still a potential "gotcha".

I'm not sure we should trust that consuming code will understand that limitation when setting in the store.  One way of handling this would be to only allow scalar values to be set via the action (so that ensures there won't be any unexpected side-effects for incoming objects/arrays). It appears that will always be the case from your implementation in this pull so we'll probably want to enforce that.